core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java

```diff
@@ -508,7 +541,17 @@ public DictionaryBlock compact()
         }
         try {
             Block compactDictionary = dictionary.copyPositions(dictionaryPositionsToCopy.elements(), 0, dictionaryPositionsToCopy.size());
-            return new DictionaryBlock(positionCount, compactDictionary, newIds, true);
+            return new DictionaryBlock(
+                    0,
+                    positionCount,
+                    compactDictionary,
+                    newIds,
+                    true,
+                    // Copied dictionary positions match ids sequence. Therefore new
+                    // compact dictionary block has sequential ids only if single position
+                    // is not used more than once.
+                    uniqueIds == positionCount,
```

Comment:
Setting it via constructor is not effective as `uniqueIds` will be `-1` and you will still call `calculateCompactSize` from `isSequentialIds`

Reply:
Added check in constructor that `isSequentialIds` cannot be set without `dictionaryIsCompacted` which happens here